File: 0135-candy/0135-candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        int t_candy = 1 ;
        int top = 0 , down = 0 , up = 0;
        for(int i = 1 ; i < n ; ++i){
            if(ratings[i] > ratings[i - 1]){
                ++up;
                down = 0;
                top = up;
                t_candy += (up + 1);
            }
            else if( ratings[i] == ratings[i - 1]){
                up = 0;
                down = 0;
                top = 0;
                ++t_candy;
            }
            else{
                up = 0;
                ++down;
                t_candy += down;
                if(down > top) ++t_candy;
            }
        }
        return t_candy;
    }
};

File: 0135-candy/0135-candy.py
class Solution:
    def candy(self, ratings: List[int]) -> int:
        candy = [1]*len(ratings)
        for i in range(1 , len(ratings)):
            if ratings[i - 1] < ratings[i]:
                candy[i] = candy[i - 1] + 1
        for i in range(len(ratings) - 2 , -1 , -1):
            if ratings[i + 1] < ratings[i] and candy[i] <= candy[i + 1]:
                candy[i] = candy[i + 1] + 1
        sum = 0
        for i in candy:
            sum += i
        return sum